robottelo/test.py

```diff
@@ -211,6 +232,128 @@ def setUp(self):
+        # For ddt tests with maps.
+        if "___" in testmethodname:
+            parts = testmethodname.split("___")
+        # For ddt tests with utf8, html, etc.
+        elif "__" in testmethodname:
+            parts = testmethodname.split("__")
+        # For ddt test with alpha, numeric, etc.
+        else:
+            testmethodname = testmethodname.split('_')
+            for _ in range(2):
+                testmethodname.pop()
+        if "___" in testmethodname or "__" in testmethodname:
+            testmethodname = parts[0]
+            testmethodname = testmethodname.split('_')
+            testmethodname.pop()
+        testmethodname = "_".join(testmethodname)
+        # Creates dir_structure depending upon the testmethodname.
+        directory = testmethodname
+        if not os.path.exists("/tmp/screenshots/" + directory):
+            os.makedirs("/tmp/screenshots/" + directory)
+        self.browser.save_screenshot('/tmp/screenshots/{dir}/'
+                                     'screenshot-{date}.png'
+                                     .format(dir=directory, date=filename))
+
+    # The method which actually runs the test methods.
+    def run(self, result=None):
+        """Overriding this method found at unittest2.case to add taking
+        screenshot support, whenever a UI Test Fails or Throws Error."""
+        orig_result = result
+        if result is None:
+            result = self.defaultTestResult()
+            startTestRun = getattr(result, 'startTestRun', None)
+            if startTestRun is not None:
+                startTestRun()
+
+        self._resultForDoCleanups = result
+        result.startTest(self)
+
+        testMethod = getattr(self, self._testMethodName)
+
+        if (getattr(self.__class__, "__unittest_skip__", False) or
+                getattr(testMethod, "__unittest_skip__", False)):
+            # If the class or method was skipped.
+            try:
+                skip_why = (getattr(self.__class__,
+                                    '__unittest_skip_why__', '')
+                            or getattr(testMethod,
+                                       '__unittest_skip_why__', ''))
+                self._addSkip(result, skip_why)
+            finally:
+                result.stopTest(self)
+            return
+        try:
+            success = False
+            try:
+                self.setUp()
+            except unittest.SkipTest, e:
+                self._addSkip(result, str(e))
+            except Exception:
+                result.addError(self, sys.exc_info())
+            else:
+                try:
+                    testMethod()
+                except self.failureException:
+                    # Takes screenshot of the UI if the test fails.
+                    self.take_screenshot(self._testMethodName)
+                    result.addFailure(self, sys.exc_info())
+                except _ExpectedFailure, e:
```

Comment:
I have copied this method run() as it is from unittest2.case. I feel we use it as it is with the needed code for adding "take screenshot" support.